### PR TITLE
Fixed #35546 -- Updated  contributing pages to align with current requirement of having an accepted ticket

### PR DESCRIPTION
**Pull request by mharyam**
# Trac ticket number
ticket-35546

# Branch description
Following some feedback from PR https://github.com/django/django/pull/17389 mentioned in the TRAC ticket, I have checked and updated the places where we need to be more explicit about the need to have an accepted ticket when working on changes.

Most of the updates involve rewording paragraphs to ensure this requirement is clearly stated.

I am happy to remove any misleading paragraphs if they are unnecessary, as we now explicitly mention the need to add a TRAC ticket for any kind of patches (whether trivial or not) in several places.

# Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 1]
```diff
@@ -2,16 +2,16 @@
 Submitting contributions
 ========================
 
-We're always grateful for contributions to Django's code. Indeed, bug reports
-with associated contributions will get fixed *far* more quickly than those
-without a solution.
+We are always grateful for patches to Django's code.
```
Comment (jacobtylerwalls) [inline docs/internals/contributing/writing-code/submitting-patches.txt:5]: A preference for the word "contributions" over "patches" was just merged in https://github.com/django/django/pull/18264, so please pull the latest and make sure this is not reverted. Thanks

File: docs/internals/contributing/bugs-and-features.txt [hunk 2]
```diff
@@ -110,10 +110,10 @@ part of that. Here are some tips on how to make a request most effectively:
 If there's a consensus agreement on the feature, then it's appropriate to
 create a ticket. Include a link to the discussion in the ticket description.
 
-As with most open-source projects, code talks. If you are willing to write the
-code for the feature yourself or, even better, if you've already written it,
-it's much more likely to be accepted. Fork Django on GitHub, create a feature
-branch, and show us your work!
+As with most open-source projects, code talks. However, before writing any code,
+it's important to follow our updated process. If you have a bug fix or feature request,
+please create a ticket first on on our `ticket tracker`_.
```
Comment (jacobtylerwalls) [inline docs/internals/contributing/bugs-and-features.txt:115]: "on on"

File: docs/internals/contributing/bugs-and-features.txt [hunk 3]
```diff
@@ -110,10 +110,10 @@ part of that. Here are some tips on how to make a request most effectively:
 If there's a consensus agreement on the feature, then it's appropriate to
 create a ticket. Include a link to the discussion in the ticket description.
 
-As with most open-source projects, code talks. If you are willing to write the
-code for the feature yourself or, even better, if you've already written it,
-it's much more likely to be accepted. Fork Django on GitHub, create a feature
-branch, and show us your work!
+As with most open-source projects, code talks. However, before writing any code,
+it's important to follow our updated process. If you have a bug fix or feature request,
```
Comment (jacobtylerwalls) [inline docs/internals/contributing/bugs-and-features.txt:114]: "Updated process" seems a bit distracting to me, since it draws attention to something changing without saying what changed. I wonder if we can replace this paragraph with a single sentence at the end of the previous one reminding folks that it is possible to either 1.) include diffs in tickets, or 2.) link to a draft PR -- the point to reinforce being that ideally the discussion starts on the ticket before moving to a PR.

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 4]
```diff
@@ -2,16 +2,16 @@
...
+To ensure a smooth process, please create a ticket on our `ticket tracker`_. before submitting your patch.
+Bug reports with associated tickets and contributions will be fixed *far* more quickly than those without.
```
Comment (jacobtylerwalls) [inline docs/internals/contributing/writing-code/submitting-patches.txt:7]: I'd avoid suggesting that bugs are occasionally fixed without tickets. Typos and minor optimizations are usually merged without one, but we don't necessarily need to mention that -- I'd keep it clear that all behavior changes, including bugfixes, need accepted tickets.

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 5]
```diff
@@ -2,16 +2,16 @@
...
 
 Typo fixes and trivial documentation changes
 ============================================
 
-If you are fixing a really trivial issue, for example changing a word in the
-documentation, the preferred way to provide the patch is using GitHub pull
-requests without a Trac ticket.
+If you are fixing a trivial issue, for example changing a word in the
+documentation, the preferred process is to create a Trac ticket first.
+Once the ticket is accepted, you can then provide the patch using a GitHub pull request.
```
Comment (jacobtylerwalls) [inline docs/internals/contributing/writing-code/submitting-patches.txt:14]: I think this part should largely stay as it was: I don't think we want to encourage tickets for typo fixes. Maybe we just need to clarify what kind of wording changes we mean. I just adjusted the wording of an error message in #18338, and I had a ticket because it was user-facing and changed the sense of the message. But if my change had only been to fix a typo, a ticket would have been unnecessary. I took the point of this section to be about _typo-like_ changes to docs that are uncontroversial in any possible way.

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 6]
```diff
@@ -270,8 +272,14 @@ appropriate flags on the Trac ticket based on the results of your review:
...
-work needs to be done. If you're looking to become a merger, doing thorough
-reviews of contributions is a great way to earn trust.
+work needs to be done.
+
+.. note::
+    Ensure that a ticket has been submitted and accepted before proceeding
+    with the patch review.
+
+If you're looking to become a merger, doing thorough reviews
+of patches is a great way to earn trust.
```
Comment (jacobtylerwalls) [inline docs/internals/contributing/writing-code/submitting-patches.txt:282]: patches -> contributions

**Review by jacobtylerwalls — COMMENTED** [review]
Thanks for picking this up! I left some suggestions.

File: docs/internals/contributing/bugs-and-features.txt (hunk 3 above)
Comment (mharyam) [reply docs/internals/contributing/bugs-and-features.txt:114]: Thank you for reviewing. 

I am unsure what updates are needed here, and I find the current paragraph a bit confusing. As I understand it, it says you can directly create a feature request,  should we clarify that you need to create a ticket on Trac before proceeding with that? Something like this:

> As with most open-source projects, code talks. If you are willing to write the
> code for the feature yourself or, even better, if you've already written it,
> it's much more likely to be accepted. Ensure you create a ticket on track first , and show us your work!

**Comment by mharyam** [comment]
@jacobtylerwalls Thanks a lot for reviewing. I have updated the changes. I just have one question left on the change request. Thank you.

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 7]
```diff
@@ -146,6 +145,8 @@ decision.
 
 If you provide a non-trivial change, include evidence that alternatives have
 been discussed on the `Django Forum`_ or |django-developers| list.
+Additionally, ensure that a ticket has been submitted and accepted before
+proceeding with your patch.
```
Comment (jacobtylerwalls) [inline docs/internals/contributing/writing-code/submitting-patches.txt:149]: The vicinity uses "contribution" rather than "patch".

File: docs/internals/contributing/bugs-and-features.txt (hunk 3 above)
Comment (jacobtylerwalls) [reply docs/internals/contributing/bugs-and-features.txt:114]: We have a few options. On Trac someone suggested just deleting the paragraph. That makes sense to me, since it does just largely repeat the generic encouragement at the top of this document, only for features instead of bugs: "bug reports with associated contributions will get fixed far more quickly than those without a solution.")

_If_ this paragraph has enough value to remain in some fashion, I think at best it reminds folks that showing a draft implementation can help others evaluate an idea. (My last comment suggested possibly reminding people to use draft PRs or embedding a diff in a ticket.) But just above this there is already "Include example code (non-functional is OK) if possible." I think people comfortable with GitHub will understand there are multiple ways to do this, with or without an actual PR, without us needing to detail it any further.

Others may have useful input here, but I'm leaning toward just deleting this paragraph. What do you think?
***
> should we clarify that you need to create a ticket on Trac

I think we might muddle things further if we just say that. The gist of the portion above this is all about encouraging people to go to the forum with any hefty feature request and waiting to open a ticket only if consensus develops.

Comment (mharyam) [reply docs/internals/contributing/bugs-and-features.txt:114]: Yes, I also agree with deleting the paragraph, as it repeats what was said before. I will go ahead and delete it.

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 8]
```diff
@@ -111,7 +111,7 @@ requirements:
   feature, the change should also contain documentation.
 
 When you think your work is ready to be reviewed, send :doc:`a GitHub pull
-request <working-with-git>`. 
+request <working-with-git>`.
```
Comment (mharyam) [inline docs/internals/contributing/writing-code/submitting-patches.txt:114]: This was automatically updated from my VSCode. I am happy to revert it back if the change is unnecessary.

File: docs/internals/contributing/bugs-and-features.txt (hunk 3 above)
Comment (mharyam) [reply docs/internals/contributing/bugs-and-features.txt]: I have updated the change requested. You can review it again. Thank you.

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 9]
```diff
@@ -270,8 +272,14 @@ appropriate flags on the Trac ticket based on the results of your review:
 "Patch needs improvement", "Needs documentation", and/or "Needs tests". As time
 and interest permits, mergers do final reviews of "Ready for checkin" tickets
 and will either commit the changes or bump it back to "Accepted" if further
-work needs to be done. If you're looking to become a merger, doing thorough
-reviews of contributions is a great way to earn trust.
+work needs to be done.
+
+.. note::
+    Ensure that a ticket has been submitted and accepted before proceeding
+    with the review.
```
Comment (jacobtylerwalls) [inline docs/internals/contributing/writing-code/submitting-patches.txt:279]: Thanks, but I can also imagine leaving this section as it was. Let's see what others think.

Comment (sarahboyce) [reply docs/internals/contributing/writing-code/submitting-patches.txt:279]: I think I agree as the second sentence onwards has the assumption that there is a ticket.

If we make a change here, I would update:
```
Use this checklist to review a pull request. If you are reviewing a pull
request that is not your own and it passes all the criteria below, please set
the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".
```
to something like
```
Use this checklist to review a pull request. If you are reviewing a non-trivial
pull request, first ensure the it has an accepted ticket. If the pull passes all 
the criteria below and is not your own, please set the "Triage Stage" on the 
corresponding Trac ticket to "Ready for checkin".
```
----

I noticed that these docs have the following sections:
1. "Typo fixes and trivial documentation changes": to me, this covers the contributions where we **do not expect a ticket**, pretty much all others we would expect an accepted ticket
2. "Non-trivial contributions": to me, this would normally mean contributions where we **expect a discussion** which concludes in an accepted ticket. However, in this context this seems to be saying that although we have accepted the ticket we still expect a discussion (whether that's the same or a new one) around the approach. We are not strict on this as a requirement, as alternative approaches are usually discussed in the review itself. This also doesn't mention DEPs and I'm not sure if that is what it is going for?

Unfortunately, that we have these two sections on the same page means that it's not very clear what we are talking about when mentioning a "non-trivial" thing.

I don't have the answer here but I would like us to link to the "Typo fixes and trivial documentation changes" section and when talking about a "non-trivial pull request" 👍 

----

Quite unrelated
> If you're looking to become a merger, doing thorough reviews of contributions is a great way to earn trust.

Perhaps "merger" might want to be updated to something like "member of the Review and Triage team" or something very generic like "hold a position of responsibility in Django". 
I appreciate that this is a separate concern to what this PR is addressing though 👍

**Comment by mharyam** [comment]
@sarahboyce I have updated the changes and linked both the trivial and non-trivial sections. This is what I have now. I think this is good, but just to confirm, do you think the heading matches and makes sense(added screenshot below)?

Asking to confirm if grammar matters, as I am linking the section title, so I have something like this: 

> **'If the change is a Non-trivial contributions,'**

 Is this okay?


Screenshot 2024-07-14 at 9 01 31 AM
Screenshot 2024-07-14 at 9 01 48 AM
Screenshot 2024-07-14 at 9 02 23 AM
Screenshot 2024-07-14 at 9 02 46 AM

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 10]
```diff
@@ -2,9 +2,11 @@
 Submitting contributions
 ========================
 
-We're always grateful for contributions to Django's code. Indeed, bug reports
-with associated contributions will get fixed *far* more quickly than those
-without a solution.
+We are always grateful for contributions to Django's code.
+To ensure a smooth process, please create a ticket for :ref:`non-trivial-contribution`
+on our `ticket tracker`_ before submitting your contribution.
```
Comment (sarahboyce) [inline docs/internals/contributing/writing-code/submitting-patches.txt:7]: I'm tempted for us to not mention tickets yet and have something like
```
Thank you for your interest in contributing to Django's code.
```
(can be bigger mentioning that Django is run on the effort of volunteers and these docs are to help you understand Django's processes and land your contribution successfully)

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 11]
```diff
@@ -52,7 +54,7 @@ and time availability), claim it by following these steps:
 
 .. note::
     The Django software foundation requests that anyone contributing more than
-    a trivial change to Django sign and submit a `Contributor License
+    a :ref:`trivial-change`, to Django sign and submit a `Contributor License
```
Comment (sarahboyce) [inline docs/internals/contributing/writing-code/submitting-patches.txt:57]: We can update the link text by doing...
```suggestion
    a :ref:`trivial change <trivial-change>`, to Django sign and submit a `Contributor License
```
(the line length will need to be adjusted)

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 12]
```diff
@@ -331,5 +341,6 @@ All tickets
   :ref:`commit message format <committing-guidelines>`?
 * Are you the patch author and a new contributor? Please add yourself to the
   :source:`AUTHORS` file and submit a `Contributor License Agreement`_.
+* If the change is a :ref:`non-trivial-contribution`, does the pull request have an accepted Trac ticket?
```
Comment (sarahboyce) [inline docs/internals/contributing/writing-code/submitting-patches.txt:344]:
```suggestion
* Does this have an accepted ticket on Trac? All contributions require a ticket
  unless the :ref:`change is considered trivial <trivial-change>`.
```

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 13]
```diff
@@ -262,16 +268,20 @@ documentation.
...
 ======================
 
-Use this checklist to review a pull request. If you are reviewing a pull
-request that is not your own and it passes all the criteria below, please set
-the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".
+Use this checklist to review to review a pull request.
+If you are reviewing a :ref:`non-trivial-contribution`,
+first ensure it has an accepted ticket.
+If the pull request passes all the criteria below and is not your own,
+please set the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".
```
Comment (sarahboyce) [inline docs/internals/contributing/writing-code/submitting-patches.txt:275]:
```suggestion
Use this checklist to review to review a pull request. If you are reviewing a
contribution that would not be :ref:`considered trivial <trivial-change>`,
first ensure it has an accepted ticket. If the pull request passes all the
criteria below and is not your own, please set the "Triage Stage" on the
corresponding Trac ticket to "Ready for checkin".
```

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 14]
```diff
@@ -262,16 +268,20 @@ documentation.
...
 appropriate flags on the Trac ticket based on the results of your review:
 "Patch needs improvement", "Needs documentation", and/or "Needs tests". As time
 and interest permits, mergers do final reviews of "Ready for checkin" tickets
 and will either commit the changes or bump it back to "Accepted" if further
-work needs to be done. If you're looking to become a merger, doing thorough
-reviews of contributions is a great way to earn trust.
+work needs to be done.
+
+If you're looking to become a member of the Review and Triage team, doing thorough reviews
+of contributions is a great way to earn trust.
```
Comment (sarahboyce) [inline docs/internals/contributing/writing-code/submitting-patches.txt:284]: Thank you for doing this, I think let's pull this out into a separate commit (or separate PR) as it more of a "side-quest" than what we're trying to achieve here 👍

**Review by sarahboyce — COMMENTED** [review]
Thank you for the updates @mharyam ⭐ 

I have updated so that we avoid linking to the "non-trivial" section and only link to the "trivial" section as the non-trivial one talks about more major changes (so doesn't include everything that we'd want a ticket for I think)

**Comment by mharyam** [comment]
@sarahboyce Thanks for reviewing, I have updated changes

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 15]
```diff
@@ -2,9 +2,12 @@
 Submitting contributions
 ========================
 
-We're always grateful for contributions to Django's code. Indeed, bug reports
-with associated contributions will get fixed *far* more quickly than those
-without a solution.
+Thank you for your interest in contributing to Django's code.
+Django is run on the effort of volunteers and these documentations
+are to help you understand Django's processes and land your contribution
+successfully.
```
Comment (sarahboyce) [inline docs/internals/contributing/writing-code/submitting-patches.txt:8]: I think we haven't got the wording of this quite right (I appreciate I suggested this wording). 

I am now thinking, this is still true. I don't think this contradicts the advice of needing a ticket. This is a bug report with a solution (ie. someone raised a ticket in trac **and** raised a pull request). Although this is not yet accepted, when someone is trying to fix a bug, it is great to see the suggested fix (especially for regressions). This happens already quite well. 

So perhaps let's leave this paragraph as it was initially. @nessita what do you think?

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 16]
```diff
@@ -146,6 +150,8 @@ decision.
 
 If you provide a non-trivial change, include evidence that alternatives have
 been discussed on the `Django Forum`_ or |django-developers| list.
+Additionally, ensure that a ticket has been submitted and accepted before
+proceeding with your contribution.
```
Comment (sarahboyce) [inline docs/internals/contributing/writing-code/submitting-patches.txt:154]: I'm tempted to say, let's drop this addition here. 

I have two reasons to think this.

Firstly, just below here it says "... ask on the ticket for opinions" which implies there is a ticket

Secondly, I am very tempted to drop this whole "Non-trivial contributions" section all together. 
This was introduced in 407f62fd31689b6158902d3ffdcde64aa27679a7 (13 years ago) and hasn't really changed since. To me, this second classification of "trivial" vs "non-trivial" within the same page makes the docs harder to follow. I also think implying that "small bug fixes" are "trivial" is a little harsh on individuals who found it quite difficult to resolve these. 

So maybe let's not touch this section for now 🤔 
@nessita do you have opinions here?

File: docs/internals/contributing/writing-code/submitting-patches.txt [hunk 17]
```diff
@@ -262,16 +268,20 @@ documentation.
...
 ======================
 
-Use this checklist to review a pull request. If you are reviewing a pull
-request that is not your own and it passes all the criteria below, please set
-the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".
+Use this checklist to review to review a pull request.
+If you are reviewing a contribution that would not be
+:ref:`considered trivial <trivial-change>`, first ensure it has an accepted ticket.
+If the pull request passes all the criteria below and is not your own,
+please set the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".
```
Comment (sarahboyce) [inline docs/internals/contributing/writing-code/submitting-patches.txt:275]: If you want a new paragraph, it needs an empty line
We also need to wrap the docs at 80 characters 👍 
```suggestion
Use this checklist to review a pull request. If this contribution would not be
:ref:`considered trivial <trivial-change>`, first ensure it has an accepted
ticket before proceeding with the review.

If the pull request passes all the criteria below and is not your own, please
set the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".
```

----

**Note:** we could also arguably avoid this change as it says:
> Use this checklist to review a pull request. If you are reviewing a pull
request that is not your own and it passes all the criteria below

and we have added the requirement of the accepted ticket to the checklist.

However, I think in this case it's worth it because we are trying to suggest a volunteer doesn't invest too much effort in reviewing something Django may have no intention of accepting

File: docs/internals/contributing/writing-code/submitting-patches.txt (hunk 14 above)
Comment (sarahboyce) [reply docs/internals/contributing/writing-code/submitting-patches.txt:284]: Thank you, this also needs wrapping at 80 characters

**Review by sarahboyce — COMMENTED** [review]
Thank you for the updates @mharyam ⭐ you're doing this really brilliantly and I appreciate your patience

I think I have added my last few "thoughts" around this change (it takes some time before I've reflected on things fully, bare with me). 
I think we're really close 👍

File: docs/internals/contributing/writing-code/submitting-patches.txt (hunk 15 above)
Comment (mharyam) [reply docs/internals/contributing/writing-code/submitting-patches.txt:8]: @sarahboyce should be this? 

> We're always grateful for contributions to Django's code. Indeed, bug reports
> with associated contributions will get fixed *far* more quickly than those
> without a solution.

Comment (sarahboyce) [reply docs/internals/contributing/writing-code/submitting-patches.txt:8]: Yes please

**Comment by mharyam** [comment]
@sarahboyce Thank you for reviewing. I have updated the changes as requested.

**Review by sarahboyce — APPROVED** [review]
Thank you @mharyam :star:
